src/freertos_drivers/pic32mx/Pic32mxUsbCdcDevice.cxx

```diff
@@ -0,0 +1,691 @@
+    /// hosts, etc.
+    inline void device_tasks();
+
+    /// This function is called from the interrupt context of the USB driver to
+    /// notify us about various events at the driver layer, such as USB plugged
+    /// in.
+    inline void event_hander_from_isr(USB_DEVICE_EVENT event, void *pData);
+
+    /// This function is called from the interrupt context of the USB driver to
+    /// notify us about various events at the CDC layer, such as TX complete.
+    inline USB_DEVICE_CDC_EVENT_RESPONSE cdc_event_hander_from_isr(
+        USB_DEVICE_CDC_EVENT event, void *pData);
+
+private:
+    static constexpr unsigned USB_CDC_BUFFER_SIZE = 64;
+
+    static constexpr unsigned TX_BUFFER_SIZE = 3 * USB_CDC_BUFFER_SIZE;
+    static constexpr unsigned RX_DEVICEBUFFER_SIZE = 0;
+
+    /// Unused.
+    void tx_char() override
+    {
+    }
+    /// Called by the serial driver when the first fd gets opened.
+    void enable() override;
+    /// Called by the serial driver when the last fd gets closed.
+    void disable() override;
+    /// Called by the serial driver after the last fd gets closed.
+    void flush_buffers() override;
+
+    /** Read from a file or device.
+     * @param file file reference for this device
+     * @param buf location to place read data
+     * @param count number of bytes to read
+     * @return number of bytes read upon success, -1 upon failure with errno
+     * containing the cause
+     */
+    ssize_t read(File *file, void *buf, size_t count) override;
+
+    /** Write to a file or device.
+     * @param file file reference for this device
+     * @param buf location to find write data
+     * @param count number of bytes to write
+     * @return number of bytes written upon success, -1 upon failure with errno
+     * containing the cause
+     */
+    ssize_t write(File *file, const void *buf, size_t count) OVERRIDE;
+
+    /** Device select method. Default impementation returns true.
+     * @param file reference to the file
+     * @param mode FREAD for read active, FWRITE for write active, 0 for
+     *        exceptions
+     * @return true if active, false if inactive
+     */
+    bool select(File *file, int mode) override;
+
+    /// Issues a read to the USB driver stack. Must be called from within a lock
+    /// or critical section. Uses next_rx_buffer().
+    void start_read_atomic();
+
+    /// Issues a write to the USB driver stack. Must be called from within a
+    /// lock or critical section or interrupt.
+    void start_write_atomic();
+
+    /// @return the side of the double buffer from which the userspace should be
+    /// reading.
+    DeviceBuffer<uint8_t> *current_rx_buffer()
+    {
+        return rxBuffers[currentReadBuffer];
+    }
+
+    /// @return the side of the double buffer into which the driver should be
+    /// writing.
+    DeviceBuffer<uint8_t> *next_rx_buffer()
+    {
+        return rxBuffers[currentReadBuffer ^ 1];
+    }
+
+    SYS_MODULE_OBJ drvUSBObject;
```

Comment:
Missing a bunch of doxygen comments.  I know I'm may be reaching, but it would be nice to not continue the practice of omitting these.

Reply:
i added them as well as i could. This is boilerplate code for the middleware to work.